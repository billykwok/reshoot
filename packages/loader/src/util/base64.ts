function base64(value: Buffer): string {
  return value.toString('base64');
}

export default base64;
