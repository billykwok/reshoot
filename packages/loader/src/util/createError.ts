import type { ErrorObject } from 'ajv';

function createError(errors: ErrorObject[]): string {
  return (
    '@reshoot/loader options are not valid:\n' +
    errors
      .map(({ dataPath, message, data }, i) => {
        if (!dataPath) {
          return `${i + 1}. options: ${message}, but got ${JSON.stringify(
            data
          )}`;
        }
        const regexResult = /^\.(.*)/gi.exec(dataPath);
        if (!regexResult) {
          return `${i + 1}. options: ${message}, but got ${JSON.stringify(
            data
          )}`;
        }
        const property = regexResult[1];
        return `${i + 1}. ${property}: ${message}, but got ${JSON.stringify(
          data
        )}`;
      })
      .join('\n')
  );
}

export default createError;
