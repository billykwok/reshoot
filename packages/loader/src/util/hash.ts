import { MetroHash64 } from 'metrohash';
import { version } from '../../package.json';

import type { AspectRatio, ResolvedOptions } from '../type';

const DELIMITER_1 = ',';
const DELIMITER_2 = '|';
const NULL = '2';
const TRUE = '1';
const FALSE = '0';

function makeBooleanHashable(value: boolean): string {
  return value ? TRUE : FALSE;
}

function makeNumberHashable(value: number): string {
  return value.toString(10);
}

function makeAspectRatioHashable(value: AspectRatio): string {
  return (
    value.type +
    DELIMITER_2 +
    value.format +
    DELIMITER_2 +
    makeNumberHashable(value.decimal)
  );
}

type SerializableOptions = Omit<ResolvedOptions, 'publicPath' | 'outputPath'>;

function makeOptionsHashable(options: SerializableOptions) {
  return (
    options.mode +
    DELIMITER_1 +
    options.name +
    DELIMITER_1 +
    options.alternativeFormats.join() +
    DELIMITER_1 +
    options.alternativeWidths.join() +
    DELIMITER_1 +
    (options.defaultFormat ? options.defaultFormat.toString() : NULL) +
    DELIMITER_1 +
    (options.defaultWidth ? makeNumberHashable(options.defaultWidth) : NULL) +
    DELIMITER_1 +
    makeNumberHashable(options.quality) +
    DELIMITER_1 +
    (options.background ? options.background : NULL) +
    DELIMITER_1 +
    (options.color ? options.color : NULL) +
    DELIMITER_1 +
    (options.placeholder
      ? makeNumberHashable(options.placeholder.size) +
        DELIMITER_2 +
        options.placeholder.trimDataUrl.toString()
      : NULL) +
    DELIMITER_1 +
    (options.aspectRatio
      ? makeAspectRatioHashable(options.aspectRatio)
      : NULL) +
    DELIMITER_1 +
    makeBooleanHashable(options.fastMode) +
    DELIMITER_1 +
    makeBooleanHashable(options.esModule)
  );
}

function computeHash(
  content: Buffer | string,
  options: SerializableOptions
): string {
  return new MetroHash64()
    .update(version + makeOptionsHashable(options))
    .update(content)
    .digest();
}

export default computeHash;
