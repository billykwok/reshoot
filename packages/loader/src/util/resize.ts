import { Mime } from '../type';

import type { Sharp } from 'sharp';
import type { ResolvedOptions } from '../type';

type Options = Pick<ResolvedOptions, 'background' | 'quality'>;

function resize(
  image: Sharp,
  width: number,
  mime: Mime,
  { background, quality }: Options
): Promise<Buffer> {
  const resized = width
    ? image.clone().resize({
        fit: 'outside',
        width: Math.max(16, width),
        height: 16,
        withoutEnlargement: true,
      })
    : image.clone();
  switch (mime) {
    case Mime.JPG:
    case Mime.JPEG:
      return resized
        .flatten(background ? { background } : false)
        .jpeg({ quality })
        .toBuffer();
    case Mime.PNG:
      return resized.png({ quality }).toBuffer();
    case Mime.WEBP:
      return resized.webp({ quality, reductionEffort: 6 }).toBuffer();
    case Mime.AVIF:
      return resized.avif({ quality }).toBuffer();
  }
  return Promise.reject(new Error(`Unsupported MIME type "${mime}"`));
}

export default resize;
