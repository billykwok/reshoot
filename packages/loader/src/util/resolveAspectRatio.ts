import { AspectRatioFormat, AspectRatioType } from '../type';

import type { AspectRatio } from '../type';

function resolveAspectRatio(
  { width, height }: { width?: number; height?: number },
  { type, format, decimal }: AspectRatio
): number {
  const scale = format === AspectRatioFormat.Percent ? 100 : 1;
  const rounder = 10 ** decimal;
  switch (type) {
    case AspectRatioType.WidthByHeight:
      return Math.floor((width / height) * scale * rounder) / rounder;
    case AspectRatioType.HeightByWidth:
      return Math.floor((height / width) * scale * rounder) / rounder;
  }
}

export default resolveAspectRatio;
