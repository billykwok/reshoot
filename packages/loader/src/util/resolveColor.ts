import color from 'color';

import type { Sharp } from 'sharp';
import { ResolvedOptions } from '../type';

async function resolveColor(
  image: Sharp,
  options: Pick<ResolvedOptions, 'background'>
): Promise<string> {
  const { dominant } = await image
    .flatten(options.background ? { background: options.background } : false)
    .stats();
  return color(dominant).hex();
}

export default resolveColor;
