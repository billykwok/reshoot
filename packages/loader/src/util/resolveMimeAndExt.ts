import path from 'path';
import { loader } from 'webpack';
import { Mime, Extension } from '../type';

function resolveMimeAndExt(
  ctx: Pick<loader.LoaderContext, 'resourcePath'>,
  enforceFormat: Mime = null
): [Mime, Extension] {
  if (enforceFormat) {
    const ext = Extension[enforceFormat];
    if (!ext) {
      throw new Error(`Format "${enforceFormat}" is not supported.`);
    }
    return [enforceFormat, Extension[enforceFormat]];
  }
  const ext = path.extname(ctx.resourcePath).replace(/\./, '') as Extension;
  const mime = Mime[ext.toUpperCase()] as Mime;
  if (!mime) {
    throw new Error(`Extension "${ext}" is not supported.`);
  }
  return [mime, ext];
}

export default resolveMimeAndExt;
