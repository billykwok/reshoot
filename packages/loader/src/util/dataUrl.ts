import base64 from './base64';

const PREFIX = 'data:image/jpeg;base64,';

export function createDataUrl(value: Buffer, trimDataUrl = false): string {
  return trimDataUrl ? base64(value) : PREFIX + base64(value);
}

export function isDataUrl(value: string): boolean {
  return /^data:/gi.test(value);
}
