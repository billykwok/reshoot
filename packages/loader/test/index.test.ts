import path from 'node:path';
import { describe, expect, test } from '@jest/globals';
import { type IFs } from 'memfs';
import { format } from 'prettier';

import { compile, createMemfs } from './util';

const IMAGE_PATH = '../../../__fixtures__/lena.png';

async function listImages(fs: IFs, pathname: string) {
  const files = (await fs.promises.readdir(
    path.join(__dirname, pathname)
  )) as string[];
  return files.filter((image) => /\.(png|jpe?g)$/i.test(image));
}

expect.addSnapshotSerializer({
  serialize: (value: string) => format(value, { parser: 'babel' }),
  test: (value: string) =>
    typeof value === 'string' &&
    /(module\.exports|export default)/i.test(value),
});

describe('Webpack with @reshoot/loader', () => {
  const options = {
    outputPath: (filename: string) =>
      `/images/${filename.replaceAll(/\/|\\/g, '_')}`,
  };

  test('should extract image meta and output image in development mode', async () => {
    const memfs = createMemfs();

    const output = await compile('development', memfs, IMAGE_PATH, options);
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './images');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image in production mode', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, options);
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './images');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image without hash length', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, {
      filename: '[contenthash].[ext]',
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image without options', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH);
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image with outputPath as a string', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, {
      outputPath: '/images/',
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './images');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image with publicPath as a function', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, {
      outputPath: '/images/',
      publicPath: (filename: string) => path.join('/images/', filename),
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './images');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image with webpack publicPath', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, options, {
      publicPath: '/static/',
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './images');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image with emitFile disabled', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, {
      emitFile: false,
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should extract image meta and output image with esModule enabled', async () => {
    const memfs = createMemfs();

    const output = await compile('production', memfs, IMAGE_PATH, {
      esModule: true,
    });
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should pass user provided color if specified', async () => {
    const memfs = createMemfs();

    const output = await compile(
      'production',
      memfs,
      `${IMAGE_PATH}?color=#eee`,
      { esModule: true }
    );
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });

  test('should decode user provided color if possible', async () => {
    const memfs = createMemfs();

    const output = await compile(
      'production',
      memfs,
      `${IMAGE_PATH}?color=%23eee`,
      { esModule: true }
    );
    expect(output).toMatchSnapshot();

    const images = await listImages(memfs, './');
    expect(images).toMatchSnapshot();

    memfs.reset();
  });
});
