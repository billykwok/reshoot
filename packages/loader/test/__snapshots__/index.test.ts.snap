// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Webpack with @reshoot/loader should extract image meta and output image in development mode 1`] = `
module.exports = {
  hash: "24793c5e741d6fa9378cf941c91bd754",
  src: "/_/_/_/__fixtures__/lena.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image in development mode 2`] = `
Array [
  "________fixtures___lena.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image in production mode 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image in production mode 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with emitFile disabled 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with emitFile disabled 2`] = `Array []`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with esModule enabled 1`] = `
export default {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with esModule enabled 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with outputPath as a string 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with outputPath as a string 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with publicPath as a function 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/images/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with publicPath as a function 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with webpack publicPath 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/static/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image with webpack publicPath 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image without hash length 1`] = `
module.exports = {
  hash: "237cd5b009a23c41445ebc847dd90cc6",
  src: "/237cd5b009a23c41.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image without hash length 2`] = `
Array [
  "237cd5b009a23c41.png",
]
`;

exports[`Webpack with @reshoot/loader should extract image meta and output image without options 1`] = `
module.exports = {
  hash: "c063cce5c8f8c5f6ab3fa5feb96ca646",
  src: "/c063cce5c8f8c5f6.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#c85858",
};

`;

exports[`Webpack with @reshoot/loader should extract image meta and output image without options 2`] = `
Array [
  "c063cce5c8f8c5f6.png",
]
`;

exports[`Webpack with @reshoot/loader should pass user provided color if specified 1`] = `
export default {
  hash: "ab38bf3f75a11f905b2211f93b2ae280",
  src: "/ab38bf3f75a11f90.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#eee",
};

`;

exports[`Webpack with @reshoot/loader should pass user provided color if specified 2`] = `
Array [
  "ab38bf3f75a11f90.png",
]
`;

exports[`Webpack with @reshoot/loader should decode user provided color if possible 1`] = `
export default {
  hash: "ab38bf3f75a11f905b2211f93b2ae280",
  src: "/ab38bf3f75a11f90.png",
  width: 512,
  height: 512,
  aspectRatio: 100,
  placeholder:
    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAIBAMAAAA2IaO4AAAA",
  color: "#eee",
};

`;

exports[`Webpack with @reshoot/loader should decode user provided color if possible 2`] = `
Array [
  "ab38bf3f75a11f90.png",
]
`;
